### PR TITLE
Add support for features

### PR DESCRIPTION
**Pull request by MikaelSmith**
Adds support for volumefs and transport features.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Example output with these new features
```
{
  "type_id": "example",
  "name": "example",
  "methods": [
    [
      "list",
      [
        {
          "type_id": "__volume::fs__",
          "name": "fs",
          "state": "{\"maxdepth\":2}"
        }
      ]
    ],
    [
      "exec",
      {
        "transport": "ssh",
        "options": {
          "host": "example.com",
          "user": "ubuntu"
        }
      }
    ],
    [
      "schema",
      {
        "example": {
          "label": "example",
          "methods": [
            "list",
            "exec",
            "schema"
          ],
          "singleton": true,
          "children": [
            "__volume::fs__"
          ]
        }
    ]
  ],
  "state": "{\"klass\":\"example\",\"name\":\"example\"}"
}
```

File: lib/wash/entry.rb [hunk 1]
```diff
@@ -343,6 +354,50 @@ def cache_ttls(ttls = {})
...
+    #
+    #     def exec
+    #       raise 'implemented by transport'
+    #     end
+    #   end
+    #
+    # @param [Symbol] klass The transport to use. Only `:ssh` is supported
+    # @param [Hash] options For SSH: :host (required), :user, :password,
+    #     :fallbackuser, :identityfile, :knownhosts, :hostkeyalias, :retries
+    def transport(klass, options = {})
```
Comment (ekinanp) [inline lib/wash/entry.rb:396]: We could define the `exec` for them and save some clutter.

File: lib/wash/entry.rb [hunk 2]
```diff
@@ -343,6 +354,50 @@ def cache_ttls(ttls = {})
       @cache_ttls = @cache_ttls.merge(ttls)
     end
 
+    # volumefs returns an entry that provides a view of the parent entry's filesystem.
+    # Use it when listing children. It can be used with `prefetch :list`.
```
Comment (ekinanp) [inline lib/wash/entry.rb]: I'd probably just say `volumefs` creates the `volume::fs` entry as specified in the external plugin docs. The current wording here makes it seem like a `wash-ruby` thing when it is more general than that.

Also, is mentioning `prefetch :list` necessary? Maybe a nit on my end, but when I read that, my brain forgets that this is just a regular entry JSON. I think your example is clear enough and shows that you can use `prefetch :list` with it. It might be nice to add one more entry in the `list` call just so "volumefs is a regular entry" is really codified

**Review by ekinanp — COMMENTED** [review]
Looks good overall, just a few comments.

File: lib/wash/entry.rb (hunk 1 above)
Comment (MikaelSmith) [reply lib/wash/entry.rb:396]: So that's challenging because the `exec` is part of the class and can be used in defining the schema, but the `transport` calls happen during instance initialization. I'd still have to have something called during class definition.

Comment (ekinanp) [reply lib/wash/entry.rb:396]: Ahh yeah you're right, we do need a class-level defn. for entry schemas so your example's correct. Would be nice to add a comment about that somewhere because it seems like that'd be a common question with a subtle answer.

Comment (MikaelSmith) [reply lib/wash/entry.rb:396]: That's right in the code comment about 12 lines up.

**Review by ekinanp — APPROVED** [review]

File: lib/wash/streamer.rb [hunk 3]
```diff
@@ -17,8 +17,8 @@ def write(chunk)
           puts("200")
           @first_chunk = false
         end
-        STDOUT.print(chunk)
-        STDOUT.flush
+        $stdout.print(chunk)
```
Comment (ekinanp) [inline lib/wash/streamer.rb:20]: Why this change?

File: spec/wash_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+require 'wash'
+
+describe Wash do
+  describe 'pretty_print?' do
```
Comment (ekinanp) [inline spec/wash_spec.rb:33]: Yeah, I was an idiot for making individual config fields helpers. It'd be nice to (at some point) refactor this into a hash that can be passed into a `Wash.config` type helper, but that's not that important right now.

File: lib/wash/streamer.rb (hunk 3 above)
Comment (MikaelSmith) [reply lib/wash/streamer.rb:20]: > Note: to_stdout and to_stderr work by temporarily replacing $stdout or $stderr, so
they're not able to intercept stream output that explicitly uses STDOUT/STDERR or that
uses a reference to $stdout/$stderr that was stored before the matcher is used.

From https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/output-matcher.

File: README.md [hunk 5]
```diff
@@ -75,6 +75,34 @@ implements `stream` and `exec`. The calling conventions and return parameters fo
...
+class Parent < Wash::Entry
+  # When 'Wash.enable_entry_schemas' is used the VOLUMEFS entry needs to be included
+  parent_of VOLUMEFS, ...
+  def list
+    [volumefs("fs", maxdepth: 2), ...]
+  end
+end
+```
+
+_Exec implementations_ provide an implementation of the `exec` method so you don't need to define your own. Wash provides an _exec implementation_ of the SSH transport. See the [external plugin exec docs](https://puppetlabs.github.io/wash/docs/external-plugins#exec) for more on _exec implementations_, including all the options available for the SSH transport. Request this transport with:
```
Comment (ekinanp) [inline README.md]: Maybe a nit, but I'd (slightly) reword this to something like

> _Exec implementations_ provide an implementation of the `exec` method so you don't need to define your own. For example if your entry works with SSH, then you can use Wash's SSH transport to implement its `exec` method. \<everything else remains the same\>

Reason: "Wash provides an _exec implementation_ of the SSH transport" feels a bit abstract to me, unless it's easy for people to figure out that that sentence means "Hey, if my thing works with SSH I can use Wash's SSH stuff to implement my exec for me"

File: lib/wash/streamer.rb (hunk 3 above)
Comment (ekinanp) [reply lib/wash/streamer.rb:20]: Ah ok, makes sense. Thanks.

**Comment by ekinanp** [comment]
I think the last commit may need to be re-worded -- it still mentions that features need to be included in the entry schema

**Review by ekinanp — APPROVED** [review]

**Comment by MikaelSmith** [comment]
I'm working on some more testing for this PR. Not very simple to add integration tests for the new Wash features, so I'm focusing on spec tests.

**Comment by MikaelSmith** [comment]
Finished.

**Review by ekinanp — APPROVED** [review]
